src/k2/transport/RPCDispatcher.cpp

```diff
@@ -150,43 +157,62 @@ void RPCDispatcher::_handleNewMessage(Request&& request) {
-            << ", with server endpoint: " << (serverep ? serverep->getURL() : String("none")));
-    if (serverep && endpoint == *serverep) {
-        // deliver via a future to possibly yield if there is a loop of send/receive requests
-        (void) seastar::sleep(0ns)
-            .then([disp=weak_from_this(), verb, endpoint, meta=std::move(meta), payload=std::move(payload)] () mutable {
-                if (disp) {
-                    // rewind the payload to the correct position
-                    payload->seek(txconstants::MAX_HEADER_SIZE);
-                    meta.setPayloadSize(payload->getDataRemaining());
-                    disp->_handleNewMessage(Request(verb, endpoint, std::move(meta), std::move(payload)));
-                }
-        });
-        return;
+            << ", with server endpoint: " << (serverep ? serverep->getURL() : String("none"))
+            << ", payload size=" << (payload?payload->getSize():0)
+            << ", payload capacity=" << (payload?payload->getCapacity():0));
+    if (_txUseCrossCoreLoopback()) {
+        if (serverep && endpoint == *serverep){
+            payload->seek(txconstants::MAX_HEADER_SIZE);
+            meta.setPayloadSize(payload->getDataRemaining());
+            _handleNewMessage(Request(verb, *RPC().getServerEndpoint(endpoint.getProtocol()), std::move(meta), std::move(payload)));
+            return seastar::make_ready_future<>();
+        }
+        
+        auto core = _url_cores.find(endpoint.getURL());
+        if (core != _url_cores.end()){
+            payload->seek(txconstants::MAX_HEADER_SIZE);
+            meta.setPayloadSize(payload->getDataRemaining());
+            auto return_value = RPCDist().invoke_on(core->second, &k2::RPCDispatcher::_handleNewMessage, Request(verb, *RPC().getServerEndpoint(endpoint.getProtocol()), std::move(meta), std::move(payload)));
+            return seastar::make_ready_future<>();
+        }
     }
     protoi->second->send(verb, std::move(payload), endpoint, std::move(meta));
+    return seastar::make_ready_future<>();
 }
 
-void RPCDispatcher::send(Verb verb, std::unique_ptr<Payload> payload, TXEndpoint& endpoint) {
+
+seastar::future<>
+RPCDispatcher::send(Verb verb, std::unique_ptr<Payload> payload, TXEndpoint& endpoint) {
     K2DEBUG("Plain send");
     MessageMetadata metadata;
-    _send(verb, std::move(payload), endpoint, std::move(metadata));
+    return _send(verb, std::move(payload), endpoint, std::move(metadata));
 }
 
-void RPCDispatcher::sendReply(std::unique_ptr<Payload> payload, Request& forRequest) {
-    K2DEBUG("Reply send for request: " << forRequest.metadata.requestID);
+seastar::future<>
+RPCDispatcher::setAddressCore(std::pair<String, int> url_core) {
+    auto url = url_core.first;
+    int core_id = url_core.second;
+    K2DEBUG("Setting address core: " << url << ", for id=" << core_id);
+    _url_cores.insert(std::make_pair(url, core_id));
```

Comment:
Can you move the url_core pair instead of reconstructing it?

Reply:
Used the move instead of reconstructing it.